File: 4.semester/ICP/src/subtractor.cpp
/**
* @file subtractor.cpp
* @brief Subtractor implementation
* @author Matúš Liščinský xlisci02, Marcin Vladimír xmarci10
*/
#include "subtractor.h"

Subtractor::Subtractor():ArithmBlock(BlockType::sub){}

/**
 * @brief Subtract two complex numbers
 */
void Subtractor::Execute(){

    std::vector<std::complex<double>> arr;
    for (Port p : this->inPorts) {
        std::complex<double> c(p.m.at("real"),p.m.at("im"));
        arr.push_back(c);
    }

    std::complex<double> result(arr[0].real(),arr[0].imag());

    for(auto &num: arr){
        if(&num == &arr[0])
            continue;
        result -= num;
    }
    this->outPorts[0].setComplex(result.real(), result.imag());
}

File: 4.semester/IJA/src/application/Subtractor.java
/* 
 * Project:     Implementing Block Editor in JavaFX
 * Course:      IJA 2018
 * Team name:   Tym xzubri00
 * Members: 
 *      Liscinsky Matus xlisci02
 *      Zubrik Tomas    xzubri00
 */
package application;

public class Subtractor extends ArithmBlock{
    
    /**
     * Constructor of Subtractor block, calls super constructor
     */
    public Subtractor() {
    	super("subtractor");
    }
    /**
     * Function subtract two complex numbers and writes the result to output port 
     */
    public void Execute() {
    	Complex num1 = new Complex(this.inPorts.get(0).m.get("real"),this.inPorts.get(0).m.get("im"));
    	Complex num2 = new Complex(this.inPorts.get(1).m.get("real"),this.inPorts.get(1).m.get("im"));
    	
    	Complex result = num1.minus(num2);
        this.outPorts.get(0).setComplex(result.re(), result.im());
    }
}

